### PR TITLE
Support refcount'ed types in numba dict

### PR DESCRIPTION
**Pull request by sklam**
(based on #3791)

as titled.

### CODE REVIEW

File: numba/tests/test_dictobject.py [hunk 1]
```diff
@@ -1019,3 +1022,189 @@ def test_repr(self):
...
+        d = TypedDict.empty(types.int64, types.unicode_type)
+        d[1] = 'apple'
+
+        @njit
+        def foo(x, k):
+            del x[1]
+
+        foo(d, 1)
+        self.assertEqual(len(d), 0)
+        self.assertFalse(d)
```
Comment (stuartarchibald) [inline numba/tests/test_dictobject.py]: Guessing that the ban on reflected types still need writing and testing?

File: numba/dictobject.py [hunk 2]
```diff
@@ -313,6 +317,154 @@ def codegen(context, builder, sig, args):
...
+        val_incref_ptr = cgutils.gep_inbounds(builder, vtable, 0, 3)
+        val_decref_ptr = cgutils.gep_inbounds(builder, vtable, 0, 4)
+
+        dm_key = context.data_model_manager[keyty.instance_type]
+        if dm_key.contains_nrt_meminfo():
+            equal = _get_equal(context, builder.module, dm_key)
+            key_incref, key_decref = _get_incref_decref(
+                context, builder.module, dm_key,
+            )
+            builder.store(
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/dictobject.py]: Non-meminfo types are POD.  The default is to do `memcmp`, which should be faster.

File: numba/_dictobject.h [hunk 3]
```diff
@@ -26,6 +40,8 @@ typedef struct {
     /* Byte offset from indices to the first entry. */
     Py_ssize_t      entry_offset;
 
+    type_based_methods_table methods;
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/_dictobject.h]: done

File: numba/_dictobject.c [hunk 4]
```diff
@@ -677,15 +723,23 @@ numba_dict_insert(
...
+        dk_incref_key(dk, key_bytes);
+        dk_incref_val(dk, val_bytes);
+
         d->used += 1;
         dk->usable -= 1;
         dk->nentries += 1;
         assert (dk->usable >= 0);
         return OK;
     }
+    /* decref old value */
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/_dictobject.c]: added comments

[the remark this answers is not in the document]
Comment (sklam) [reply numba/_dictobject.c]: and the else-block

File: numba/tests/test_dictobject.py [hunk 5]
```diff
@@ -819,39 +823,38 @@ def foo():
...
-            for x in d.items():
-                print(x)
-
-        print(foo())
+            out = list(d.items())
+            return out
+        self.assertEqual(foo(), [('a' * 10000, 1)])
 
     def test_022_references_juggle(self):
         # this should work, llvmlite level broken, probably the same problem as
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_dictobject.py]: removed

File: numba/tests/test_dictobject.py [hunk 6]
```diff
@@ -1019,3 +1022,189 @@ def test_repr(self):
...
+        self.assertFalse('velocity' in d)
+        del expect[0:2]
+
+        for i in range(90):
+            k, v = d.popitem()
+            w = expect.pop()
+            self.assertPreciseEqual(v, w)
+
+        # Trigger a resize
+        expect.append(np.random.random(10))
```
[the remark this answers is not in the document]
Comment (sklam) [reply numba/tests/test_dictobject.py]: added seeding

File: numba/tests/test_dictobject.py (hunk 1 above)
Comment (sklam) [reply numba/tests/test_dictobject.py]: right, still need to do this

Comment (sklam) [reply numba/tests/test_dictobject.py]: done

**Comment by sklam** [comment]
Launch CPU smoketest on the farm

**Pull request merged**